pkg/util/httpgrpcutil/middleware/grpc_instrumentation.go

```diff
@@ -0,0 +1,116 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package middleware
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+
+	"github.com/gorilla/mux"
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/server"
+	"google.golang.org/grpc"
+)
+
+// InitGRPCMiddleware initializes the stuff
+func InitGRPCMiddleware(cfg *server.Config) {
+	cfg.GRPCMiddleware = append(cfg.GRPCMiddleware, OpenTracingHTTPGRPCUnaryServerInterceptor(opentracing.GlobalTracer()))
+}
+
+const httpGRPCHandleMethod = "/httpgrpc.HTTP/Handle"
+const httpSpanNameSep = " "
+
+// InitHTTPMiddleware initializes the stuff
+func InitHTTPMiddleware(router *mux.Router) {
+	middleware := OpenTracingHTTPGRPCMiddleware{router: router}
+
+	//middlewareFunc := func(handler http.Handler) http.Handler {
+	//	return nethttp.Middleware(opentracing.GlobalTracer(), router)
+	//}
+	//
+	//router.Use(middlewareFunc)
+	router.Use(middleware.Wrap)
+}
+
+type OpenTracingHTTPGRPCMiddleware struct {
+	router *mux.Router
+}
+
+func (m *OpenTracingHTTPGRPCMiddleware) Wrap(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		tracer := opentracing.GlobalTracer()
+
+		matchedRoute, _ := mux.CurrentRoute(r).GetPathTemplate()
+
+		parentSpan := opentracing.SpanFromContext(ctx)
+		parentSpan.SetTag("http.request.method", r.Method)
+		parentSpan.SetTag("http.route", matchedRoute)
+		parentSpan.SetTag("url.path", r.RequestURI)
+
+		childSpanName := r.Method + httpSpanNameSep + matchedRoute
+		startSpanOpts := []opentracing.StartSpanOption{
+			ext.SpanKindRPCServer,
+			opentracing.Tag{Key: string(ext.Component), Value: "net/http"},
+			opentracing.Tag{Key: "http.request.method", Value: r.Method},
+			opentracing.Tag{Key: "http.route", Value: matchedRoute},
+			opentracing.Tag{Key: "url.path", Value: r.RequestURI},
+		}
+		childSpan, _ := opentracing.StartSpanFromContextWithTracer(ctx, tracer, childSpanName, startSpanOpts...)
+		defer childSpan.Finish()
+
+		next.ServeHTTP(w, r)
```

Comment:
Do we need to do something here to pass the new context (with the new `childSpan` span) to `r`, so that spans created by `next` are nested under `childSpan`? Based on the screenshot you shared, it looks like child spans are nested under `parentSpan`, rather than `childSpan`.

Reply:
I also noticed this and I'm going to poke at it a bit